invenio_vocabularies/contrib/subjects/hierarchal_subject_vocabularies/datastreams.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2024 University of Münster.
+#
+# Invenio-Vocabularies is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Names datastreams, transformers, writers and readers."""
+
+import io
+
+import requests
+from rdflib import RDF, Graph, Namespace
+
+from invenio_vocabularies.datastreams.readers import BaseReader
+from invenio_vocabularies.datastreams.transformers import BaseTransformer
+
+
+class BaseHierarchalSubjectsHTTPReader(BaseReader):
+    """Base Reader class to fetch and process RDF data."""
+
+    SKOS_CORE = Namespace("http://www.w3.org/2004/02/skos/core#")
+
+    def __init__(self, origin=None, mode="r", since=None, *args, **kwargs):
+        """Initialize the reader with the data source.
+
+        :param origin: The URL from which to fetch the RDF data.
+        :param mode: Mode of operation (default is 'r' for reading).
+        """
+        self.origin = origin
+        super().__init__(origin=origin, mode=mode, *args, **kwargs)
+
+    def _iter(self, rdf_graph):
+        """Iterate over the RDF graph, yielding one subject at a time."""
+        for subject, _, _ in rdf_graph.triples(
+            (None, RDF.type, self.SKOS_CORE.Concept)
+        ):
+            yield {"subject": subject, "rdf_graph": rdf_graph}
+
+    def read(self, item=None, *args, **kwargs):
+        """Fetch and process the RDF data, yielding it one subject at a time."""
+        if item:
+            raise NotImplementedError(
+                f"{self.__class__.__name__} does not support being chained after another reader"
+            )
+
+        response = requests.get(self.origin)
+        response.raise_for_status()
+
+        rdf_content = self._get_rdf_content(response)
+        rdf_graph = Graph()
+        rdf_graph.parse(io.StringIO(rdf_content), format="xml")
+
+        yield from self._iter(rdf_graph)
+
+    def _get_rdf_content(self, response):
+        """Get RDF content from the HTTP response."""
+        return response.content.decode("utf-8")
+
+
+class BaseHierarchalSubjectsTransformer(BaseTransformer):
+    """Base Transformer class for RDF data to dictionary format."""
+
+    SKOS_CORE = Namespace("http://www.w3.org/2004/02/skos/core#")
+    SPLITCHAR = ","
+
+    def _get_labels(self, subject, rdf_graph):
+        """Extract labels (prefLabel or altLabel) for a subject."""
+        labels = {
+            label.language: label.value.capitalize()
+            for _, _, label in rdf_graph.triples(
+                (subject, self.SKOS_CORE.prefLabel, None)
+            )
+            if label.language and "-" not in label.language
+        }
+
+        if "en" not in labels:
+            for _, _, label in rdf_graph.triples(
+                (subject, self.SKOS_CORE.altLabel, None)
+            ):
+                labels.setdefault(label.language, label.value.capitalize())
```

Comment:
Does this guarantee we will have a `en` label?

Reply:
for euroscivoc the alternaive label was always in eng, but gemet has more language supported, so yeah would need to add the check

Reply:
the labels would have all languages,  if `en` is not found in preferred labels, it will try to find it in alternative labels and then the labels are sent down to the child class, where we extract the eng label for subject while rest are added to titles
```
        return {
            "id": id,
            "scheme": "EuroSciVoc",
            "subject": labels.get("en", "").capitalize(),
            "title": labels,
            "props": {"parents": parents} if parents else {},
            "identifiers": identifiers,
        }
```